GraphqlService has no env-token path: a credential-less Data Sync stage fails with `gh auth login` advice on CI

> **RE-SCOPED 2026-07-26 after @neo-gpt-emmy's cycle-1 review of PR #15988** (`Close-Target Audit: Fail`). This ticket originally bundled the credential *resolver* fix with the credential *scope* decision and pipeline recovery. PR #15988 delivers only the resolver, so per close-target remediation this ticket is narrowed to the leaf that PR fully delivers. The remaining work is named in **Split Out** below and stays open under explicit authority rather than being closed by implication.

## Context

The Data Sync Pipeline has been failing on `dev` since **2026-07-17** — **98 of the last 100 `dev` runs**, all scheduled (watchdog alarm `#15972`, which states it "only carries the alarm" and cannot host root-cause work; that alarm reports the streak as **4**, because it is not branch-scoped — separately owned by @neo-opus-ada). Root cause traced to the proving line: run `30201476105`, step `Run bounded Data Sync emission and publish`.

```
[DataSync] emit attempt=1 stage=content indexes and SEO credential=none
[ERROR] Failed to get GitHub auth token from `gh` CLI. Error: Command failed: gh auth token
Error: Could not authenticate with GitHub. Please ensure you have run `gh auth login`.
  at #getAuthToken (ai/services/github-workflow/GraphqlService.mjs:114)
  at LabelService.listLabels:58 → createLabelIndex:59 → rebuildContentIndexesAndSeo:67
```

> **CAUSAL FRAMING CORRECTED 2026-07-26 (third pass), on @neo-opus-ada's re-measurement.** This ticket first called #15744 the *regression*, then the *revealer*. **Both were wrong, and both were downstream of one bad query.** I measured with `gh run list --workflow=data-sync-pipeline.yml` and **no `--branch` filter**, so the "last success" I anchored a regression window on — `2026-07-26T00:17:01Z` — was a **feature-branch dispatch** (`agent/15744-data-sync-app-identity`, `event=workflow_dispatch`), not a `dev` run. There was no window.

**The real shape: one continuous outage beginning 2026-07-17, never recovered on `dev`.** Verified independently at `repos/neomjs/neo/actions/workflows/data-sync-pipeline.yml/runs?branch=dev&per_page=100`:

- **98 failures / 2 successes** in the last 100 `dev` runs.
- Last `dev` success: **`2026-07-17T03:20:56Z`** (run `29552162329`, `event=schedule`).
- Last corpus-facet commit: **`2026-07-17T05:13:29Z`** — *two hours later, same day*.

Those are not two coincidental lapses. The corpus went stale **because** the pipeline stopped succeeding. And #15744 landed `2026-07-26T01:55:33Z` — **nine days into** the outage — so it is neither the regression nor the revealer of it. Whatever #15744 fixed was validated on its own branch (the same run polluting the watchdog streak) and never held on `dev`.

**What survives unchanged:** `scopedStageEnv`'s JSDoc still declares the credential-less failure correct by design — *"a stage that turns out to need one fails loudly on its own missing-auth path rather than quietly succeeding on a more privileged identity than it was granted."* The `credential=none` mis-declaration is real and is the defect; it simply predates #15744 by at least nine days rather than being exposed by it. Nothing here loosens #15744, and its permission work stands on its own merits.

## The Problem (this leaf only)

`GraphqlService#getAuthToken` resolved exactly three ways: `authTokenOverride` → cached `#authToken` → `execAsync('gh auth token')`. **No environment path existed.** Consequences:

1. Every CI consumer of the service silently depended on an authenticated `gh` CLI on the runner.
2. A missing credential surfaced as `no oauth token found for github.com` plus advice to run `gh auth login` — on a GitHub Actions runner, where that is meaningless. The design *depends* on a credential-less stage "failing loudly on its own missing-auth path"; that path was loud but **misdirecting**, which is what let it run four scheduled failures deep.

## Contract Ledger

| Target surface | Source of authority | Required behaviour | Failure mode | Evidence |
|---|---|---|---|---|
| `GraphqlService#getAuthToken` | this leaf | resolve `authTokenOverride` → `GH_TOKEN` → `GITHUB_TOKEN` → cached → `gh auth token` | no credential ⇒ error naming the env vars | unit precedence suite |
| Env-vs-cache asymmetry | cost + staleness, **not** isolation | env read every call; only the CLI result memoized | a memoized env value serves a rotated credential to long-lived in-process consumers | JSDoc rationale + review-confirmed process topology |
| Stage failure legibility | `dataSyncPipeline` scope annotation | a failing stage names its declared `tokenScope` | bare child failure reads as "the tool is broken" | annotated `catch` at the emission loop |
| Per-child credential isolation | `scopedStageEnv()` | unchanged — both vars stripped before spawn | a `none`-scoped stage must read nothing | unchanged code + review-confirmed |
| Documented contract | ledger-is-a-spec | class summary, `restApiUrl`, `authTokenOverride` docs match behaviour | a stale doc contradicts the code inverted | three doc sites corrected |

## Acceptance Criteria

- [ ] `#getAuthToken` resolves override → `GH_TOKEN` → `GITHUB_TOKEN` → cached → `gh auth token`, with the no-credential error naming the env vars instead of `gh auth login`.
- [ ] Unit coverage for the override and environment branches **plus** the no-credential error branch through a deterministic CLI stand-in.
- [ ] The env-vs-cache rationale is **process-correct**: justified by cost and the staleness window for long-lived in-process consumers, never by cross-stage leakage (a fresh child process per stage makes singleton leakage impossible there).
- [ ] Stage failures carry their declared `tokenScope`; no second hand-maintained per-stage requirement declaration is introduced.
- [ ] `scopedStageEnv()` isolation provably unchanged.
- [ ] The three stale doc sites (class summary, `restApiUrl`, `authTokenOverride`) match the new behaviour.

## Split Out — bound to a named open successor

> **BOUND 2026-07-26 (cycle-2 close-target remediation).** @neo-gpt-emmy's audit was correct that a "remains open" section inside an **auto-closing** leaf is not an authority surface, and that a closed ticket (`#15744`) and an alarm-only issue (`#15972`) cannot be successors. **`#15993` is now that surface**, and it carries items 1–2 with the full evidence, the three candidate shapes, and their trades. This leaf may therefore close without orphaning the residual.

1. **The credential-scope fork → `#15993`.** The label-index stage needs a repository-scoped read credential and no existing identity fits: `intake` is installed on the DevIndex repos, `publisher` holds `contents: write` where labels need `issues`/metadata, and the implicit token would be a **third** credential in a workflow whose comments document a deliberate reduction to two. Those options differ in least-privilege posture, so the decision belongs to whoever owns **#15744**'s split — explicitly not made here, and `#15993` records all three shapes rather than pre-selecting one.
2. **Pipeline recovery / green scheduled run → `#15993`.** Blocked on (1). The pipeline stays red after this leaf **by design** — the design is refusing to run a stage under-privileged, and this leaf makes the refusal legible rather than granting the privilege. `#15972` therefore will **not** auto-close on PR #15988; @neo-opus-ada reached that independently. `#15993` carries the green-scheduled-run AC, including that a `workflow_dispatch` on a feature branch does not count.
3. **A credential-axis preflight — WITHDRAWN, not deferred, so it needs no successor.** Predicting which stages require a credential needs a second hand-maintained declaration beside `tokenScope`, free to drift from it, with nothing deriving either from what the stage does — and whether a command needs auth is not statically derivable from its argv. PR #15988 delivers post-failure annotation instead and must not be read as delivering a preflight. Withdrawn work is not residual work; a genuine preflight would need a *derived* requirement, which is a new design question rather than an unfinished obligation.

## Out of Scope

- **#15977** (generator scheduled / publisher unowned) — different defect.
- **#15963 / PR #15964** — lands the ISSUE facet's *data*; complementary, not this.
- Restoring `contents: write`, widening either App installation, or adding a secret.

## Related

`#15972` (watchdog alarm) · `#15744` / `b901103ae5` (the revealer; its permission work stands) · `#15977` · `#15963` · `#15751`.

Origin Session ID: 0b42f11c-b322-4387-8add-e4922717ff76

Authored by Grace (@neo-opus-grace, Claude Opus 5, Claude Code).

--- a/ai/services/github-workflow/GraphqlService.mjs
+++ b/ai/services/github-workflow/GraphqlService.mjs
@@ -9,9 +9,9 @@ const execAsync = promisify(exec);
  * @summary A centralized, singleton service for interacting with the GitHub GraphQL API.
  *
  * This service encapsulates all the logic for making authenticated GraphQL queries and mutations.
- * It handles fetching the auth token from the `gh` CLI, caching it, and attaching it to
- * all outgoing requests. It also provides a generic `query` method for executing
- * GraphQL operations and basic error handling.
+ * It resolves the credential (explicit override → `GH_TOKEN` → `GITHUB_TOKEN` → cached `gh auth
+ * token`; see {@link #getAuthToken}), attaches it to all outgoing requests, and provides a generic
+ * `query` method for executing GraphQL operations plus basic error handling.
  *
  * @class Neo.ai.services.github-workflow.GraphqlService
  * @extends Neo.core.Base
@@ -36,7 +36,7 @@ class GraphqlService extends Base {
          */
         apiUrl: 'https://api.github.com/graphql',
         /**
-         * The GitHub REST API base URL. REST requests share this service's cached `gh auth token`
+         * The GitHub REST API base URL. REST requests share this service's resolved credential
          * and retry machinery (see {@link #rest}), so REST writes run through the same
          * authenticated, retry-equipped path as GraphQL instead of a fresh per-call `spawn('gh')`.
          * @member {String} restApiUrl='https://api.github.com'
@@ -44,8 +44,9 @@ class GraphqlService extends Base {
          */
         restApiUrl: 'https://api.github.com',
         /**
-         * Optional explicit token override for tests or controlled embedded callers.
-         * Normal runtime auth continues to use `gh auth token`.
+         * Optional explicit token override for tests or controlled embedded callers. Outranks every
+         * other source. Absent it, runtime auth reads `GH_TOKEN`/`GITHUB_TOKEN` and falls back to
+         * `gh auth token` for interactive use.
          * @member {String|null} authTokenOverride=null
          * @protected
          */
@@ -90,15 +91,42 @@ class GraphqlService extends Base {
     #authToken = null;
 
     /**
-     * Fetches the GitHub authentication token from the `gh` CLI and caches it.
-     * This is the only method in the service that should interact with the `gh` CLI.
-     * @returns {Promise<string>} The authentication token.
-     * @throws {Error} If the token cannot be fetched.
+     * @summary Resolves the GitHub credential, preferring the ambient environment over the `gh` CLI.
+     *
+     * Precedence: explicit override → `GH_TOKEN` → `GITHUB_TOKEN` → cached CLI token → `gh auth token`.
+     * This is the only method in the service that interacts with the `gh` CLI.
+     *
+     * **Why the env vars come before the CLI.** The CLI shell-out is an interactive-developer
+     * affordance, and it was once the *only* path here. CI runners have no `gh` login, so a missing
+     * credential surfaced as `no oauth token found for github.com` plus advice to run `gh auth login`
+     * — on a machine where that is meaningless, which turned a missing-credential condition into an
+     * apparent tooling fault and hid it through repeated scheduled failures. Reading the environment
+     * first makes that failure mode structurally impossible for every CI consumer of this service,
+     * not just the caller that surfaced it.
+     *
+     * **Why the env value is deliberately NOT cached.** `process.env` is already the cache, and caching
+     * it would defeat mid-process rotation: `buildScripts/dataSyncPipeline.mjs`'s `scopedStageEnv()`
+     * rewrites `GH_TOKEN`/`GITHUB_TOKEN` per stage precisely so an intake stage cannot read the
+     * publisher credential. A cached first-stage token would silently serve the wrong identity to every
+     * later stage — the additive-boundary defect that scoping exists to prevent. Only the CLI result is
+     * cached, because shelling out per call is the cost caching was introduced for.
+     *
+     * **Invariant:** the raw token is never logged. On failure the message names the env vars to set,
+     * because the previous message sent CI operators down an interactive path that cannot exist there.
+     *
+     * @returns {Promise<String>} The authentication token.
+     * @throws {Error} When no credential is available from the override, the environment, or the CLI.
      * @private
      */
     async #getAuthToken() {
         if (this.authTokenOverride) {
             return String(this.authTokenOverride).trim();
+        }
+
+        const envToken = process.env.GH_TOKEN?.trim() || process.env.GITHUB_TOKEN?.trim();
+
+        if (envToken) {
+            return envToken;
         }
 
         if (this.#authToken) {
@@ -110,8 +138,8 @@ class GraphqlService extends Base {
             this.#authToken = stdout.trim();
             return this.#authToken;
         } catch (e) {
-            logger.error('Failed to get GitHub auth token from `gh` CLI.', e);
-            throw new Error('Could not authenticate with GitHub. Please ensure you have run `gh auth login`.');
+            logger.error('Failed to get a GitHub token: neither GH_TOKEN nor GITHUB_TOKEN is set, and the `gh` CLI holds no token.', e);
+            throw new Error('Could not authenticate with GitHub. Set GH_TOKEN or GITHUB_TOKEN (CI), or run `gh auth login` (interactive).');
         }
     }
 

--- a/ai/services/github-workflow/GraphqlService.mjs
+++ b/ai/services/github-workflow/GraphqlService.mjs
@@ -104,12 +104,18 @@ class GraphqlService extends Base {
      * first makes that failure mode structurally impossible for every CI consumer of this service,
      * not just the caller that surfaced it.
      *
-     * **Why the env value is deliberately NOT cached.** `process.env` is already the cache, and caching
-     * it would defeat mid-process rotation: `buildScripts/dataSyncPipeline.mjs`'s `scopedStageEnv()`
-     * rewrites `GH_TOKEN`/`GITHUB_TOKEN` per stage precisely so an intake stage cannot read the
-     * publisher credential. A cached first-stage token would silently serve the wrong identity to every
-     * later stage — the additive-boundary defect that scoping exists to prevent. Only the CLI result is
-     * cached, because shelling out per call is the cost caching was introduced for.
+     * **Why the env value is deliberately NOT cached.** `process.env` is already the cache — re-reading
+     * it costs nothing, so memoizing it buys no performance while adding a staleness window. The
+     * asymmetry with the CLI branch is therefore about cost, not safety: shelling out is expensive, so
+     * that result is cached; an env read is not, so it is not.
+     *
+     * The staleness window matters for **long-lived in-process consumers** whose credential can be
+     * re-pointed between calls — an embedded caller swapping identity, or a rotated token in a
+     * persistent server. A memoized env value would keep serving the pre-rotation credential for the
+     * lifetime of the process. This is NOT justified by the per-stage scoping in
+     * `buildScripts/dataSyncPipeline.mjs`: that pipeline spawns a **fresh child process per stage**
+     * (`executeCommand()`), so a singleton cache cannot cross stages there and its isolation does not
+     * depend on this choice.
      *
      * **Invariant:** the raw token is never logged. On failure the message names the env vars to set,
      * because the previous message sent CI operators down an interactive path that cannot exist there.

--- a/buildScripts/dataSyncPipeline.mjs
+++ b/buildScripts/dataSyncPipeline.mjs
@@ -430,7 +430,26 @@ export async function emitGeneratedData({
 
     for (const {args, command, label, tokenScope} of emissionCommands) {
         log(`[DataSync] emit attempt=${attempt} stage=${label} credential=${tokenScope}`);
-        await execute(command, args, {cwd, env: scopedStageEnv(tokenScope)})
+
+        try {
+            await execute(command, args, {cwd, env: scopedStageEnv(tokenScope)})
+        } catch (error) {
+            // The scope annotation is the only thing that knows which identity this stage was
+            // entitled to; the failing child does not. So a bare child failure reads as "the tool
+            // is broken" when the finding is "this stage was granted `none` and needs a credential".
+            // That misreading has already cost several scheduled runs, because a child's own
+            // missing-auth message can advise an interactive login that CI cannot perform.
+            //
+            // Deliberately NOT a per-stage `requiresCredential` flag: that would be a second
+            // hand-maintained declaration beside `tokenScope`, free to drift from it, with nothing
+            // deriving either from what the stage actually does. Annotating the failure is derived
+            // from observed behaviour and cannot disagree with itself.
+            error.message = `[DataSync] stage "${label}" failed under declared credential scope ` +
+                `\`${tokenScope}\`. If this is an authentication failure, the stage requires a scope ` +
+                `that grants it — never an ambient credential.\n${error.message}`;
+
+            throw error
+        }
     }
 }
 

--- a/test/playwright/unit/ai/buildScripts/DataSyncPipeline.spec.mjs
+++ b/test/playwright/unit/ai/buildScripts/DataSyncPipeline.spec.mjs
@@ -406,6 +406,28 @@ test.describe('tokenScope validation fails closed', () => {
             preflight: async () => {}
         })).resolves.toBeUndefined()
     });
+
+    test('a failing stage names the scope it was granted, not just the child error', async () => {
+        // A bare child failure reads as "the tool is broken" when the finding is "this stage was
+        // granted `none` and needs a credential". That misreading cost a nine-day outage: the
+        // child's own missing-auth message advised an interactive login CI cannot perform, so the
+        // declared-scope context is the part that makes the failure diagnosable at all.
+        //
+        // The annotation must therefore carry BOTH the stage label and its declared scope --
+        // asserting only that the original error survives would pass without the annotation.
+        const failure = new Error('child exited with code 1');
+
+        await expect(emitGeneratedData({
+            attempt  : 1,
+            cwd      : '/tmp',
+            execute  : async () => { throw failure },
+            log      : () => {},
+            preflight: async () => {}
+        })).rejects.toThrow(/stage "install dependencies" failed under declared credential scope `none`/);
+
+        // the child's own message is preserved, not replaced -- the annotation prefixes context
+        expect(failure.message).toContain('child exited with code 1')
+    });
 });
 
 /**

--- a/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
+++ b/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
@@ -366,3 +366,139 @@ test.describe('Neo.ai.services.github-workflow.GraphqlService — rest() authent
         expect(callCount).toBe(1);
     });
 });
+
+/**
+ * Credential resolution order.
+ *
+ * The only source `#getAuthToken` once consulted was `gh auth token`, and **every describe block
+ * above sets `authTokenOverride`** — so the resolution path itself was never exercised. That is how
+ * a CI consumer could depend on an authenticated `gh` CLI unnoticed until scheduled runs failed on
+ * it, reporting a missing credential as advice to run an interactive login CI cannot perform.
+ *
+ * These pin the order: override → `GH_TOKEN` → `GITHUB_TOKEN` → cached → CLI.
+ *
+ * The CLI branch is deliberately NOT asserted. `gh auth token` succeeds on a developer machine and
+ * fails on CI, so a test touching it would pin the environment rather than the code — the shape of
+ * false green this suite exists to avoid. Env-before-cache is likewise intentional and load-bearing:
+ * `buildScripts/dataSyncPipeline.mjs` rewrites these variables per stage so an intake stage cannot
+ * read the publisher credential, and a cached first-stage token would silently serve the wrong
+ * identity to every later stage.
+ */
+test.describe('Neo.ai.services.github-workflow.GraphqlService — credential resolution order (#15986)', () => {
+    let GraphqlService;
+    let originalAuthTokenOverride;
+    let originalFetch;
+    let originalGhToken;
+    let originalGithubToken;
+
+    const QUERY = 'query TestQuery { viewer { login } }';
+
+    /**
+     * Stubs `fetch` and captures the Authorization header the service actually sent.
+     * @returns {Object} A holder whose `value` is populated by the next call.
+     */
+    const captureAuthHeader = () => {
+        const seen = {value: null};
+
+        globalThis.fetch = async (url, options) => {
+            seen.value = new Headers(options.headers).get('authorization');
+
+            return new Response(JSON.stringify({data: {viewer: {login: 'neo-opus-grace'}}}), {
+                status : 200,
+                headers: {'content-type': 'application/json'}
+            });
+        };
+
+        return seen
+    };
+
+    test.beforeAll(async () => {
+        GraphqlService = (await import('../../../../../../ai/services/github-workflow/GraphqlService.mjs')).default;
+    });
+
+    test.beforeEach(() => {
+        originalAuthTokenOverride = GraphqlService.authTokenOverride;
+        originalFetch             = globalThis.fetch;
+        originalGhToken           = process.env.GH_TOKEN;
+        originalGithubToken       = process.env.GITHUB_TOKEN;
+
+        GraphqlService.authTokenOverride = null;
+
+        delete process.env.GH_TOKEN;
+        delete process.env.GITHUB_TOKEN;
+    });
+
+    test.afterEach(() => {
+        globalThis.fetch = originalFetch;
+
+        GraphqlService.authTokenOverride = originalAuthTokenOverride;
+
+        if (originalGhToken === undefined) {
+            delete process.env.GH_TOKEN;
+        } else {
+            process.env.GH_TOKEN = originalGhToken;
+        }
+
+        if (originalGithubToken === undefined) {
+            delete process.env.GITHUB_TOKEN;
+        } else {
+            process.env.GITHUB_TOKEN = originalGithubToken;
+        }
+    });
+
+    test('reads GH_TOKEN from the environment when no override is set', async () => {
+        process.env.GH_TOKEN = 'env-gh-token';
+
+        const seen = captureAuthHeader();
+
+        await GraphqlService.query(QUERY);
+
+        expect(seen.value).toContain('env-gh-token');
+    });
+
+    test('falls back to GITHUB_TOKEN when GH_TOKEN is absent', async () => {
+        process.env.GITHUB_TOKEN = 'env-github-token';
+
+        const seen = captureAuthHeader();
+
+        await GraphqlService.query(QUERY);
+
+        expect(seen.value).toContain('env-github-token');
+    });
+
+    test('prefers GH_TOKEN over GITHUB_TOKEN when both are set', async () => {
+        process.env.GH_TOKEN     = 'env-gh-token';
+        process.env.GITHUB_TOKEN = 'env-github-token';
+
+        const seen = captureAuthHeader();
+
+        await GraphqlService.query(QUERY);
+
+        expect(seen.value).toContain('env-gh-token');
+        expect(seen.value).not.toContain('env-github-token');
+    });
+
+    test('an explicit override still outranks the environment', async () => {
+        GraphqlService.authTokenOverride = 'override-token';
+        process.env.GH_TOKEN             = 'env-gh-token';
+
+        const seen = captureAuthHeader();
+
+        await GraphqlService.query(QUERY);
+
+        expect(seen.value).toContain('override-token');
+        expect(seen.value).not.toContain('env-gh-token');
+    });
+
+    test('treats a whitespace-only credential as absent rather than sending it', async () => {
+        process.env.GH_TOKEN     = '   ';
+        process.env.GITHUB_TOKEN = 'env-github-token';
+
+        const seen = captureAuthHeader();
+
+        await GraphqlService.query(QUERY);
+
+        expect(seen.value).toContain('env-github-token');
+        expect(seen.value).not.toContain('   ');
+    });
+});

--- a/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
+++ b/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
@@ -375,14 +375,18 @@ test.describe('Neo.ai.services.github-workflow.GraphqlService — rest() authent
  * a CI consumer could depend on an authenticated `gh` CLI unnoticed until scheduled runs failed on
  * it, reporting a missing credential as advice to run an interactive login CI cannot perform.
  *
- * These pin the order: override → `GH_TOKEN` → `GITHUB_TOKEN` → cached → CLI.
+ * These cover the override and environment branches: override → `GH_TOKEN` → `GITHUB_TOKEN`.
  *
- * The CLI branch is deliberately NOT asserted. `gh auth token` succeeds on a developer machine and
- * fails on CI, so a test touching it would pin the environment rather than the code — the shape of
- * false green this suite exists to avoid. Env-before-cache is likewise intentional and load-bearing:
- * `buildScripts/dataSyncPipeline.mjs` rewrites these variables per stage so an intake stage cannot
- * read the publisher credential, and a cached first-stage token would silently serve the wrong
- * identity to every later stage.
+ * **Scope limit, stated rather than implied:** the cached-CLI and CLI-shell-out branches are NOT
+ * covered here, so this suite pins the *environment* precedence, not the full resolution order. A
+ * bare `gh auth token` assertion succeeds on a developer machine and fails on CI, which would pin the
+ * environment instead of the code. Covering them needs a deterministic CLI seam.
+ *
+ * Env-before-cache is intentional, but for **cost and staleness**, not isolation: an env read is free
+ * so memoizing it buys nothing while adding a staleness window for long-lived in-process consumers
+ * whose credential is re-pointed between calls. It is *not* justified by `dataSyncPipeline`'s
+ * per-stage scoping — that pipeline spawns a fresh child process per stage, so a singleton cache
+ * cannot cross stages there at all.
  */
 test.describe('Neo.ai.services.github-workflow.GraphqlService — credential resolution order (#15986)', () => {
     let GraphqlService;
@@ -500,5 +504,41 @@ test.describe('Neo.ai.services.github-workflow.GraphqlService — credential res
 
         expect(seen.value).toContain('env-github-token');
         expect(seen.value).not.toContain('   ');
+    });
+
+    /**
+     * The no-credential error path, driven through a deterministic `gh` stand-in.
+     *
+     * A `PATH`-prepended stub makes the CLI branch fail on demand without depending on whether the
+     * host has a real `gh` login — the same seam used to red-prove this change, and the reason the
+     * red proof could not simply unset the token: on an authenticated developer machine the real CLI
+     * would have returned a live credential into the assertion diff.
+     *
+     * **Only the failing CLI branch is exercised here, deliberately.** A *successful* CLI call
+     * populates the service's private `#authToken` cache, which has no reset seam, so asserting it
+     * would leave a resolved credential on the singleton and silently change what every later test in
+     * the worker resolves — order-dependent pollution. The failure path caches nothing, so it is
+     * order-independent. Covering the cached and CLI-success branches needs a cache-reset seam on the
+     * service; that is a deliberate API cost and is left unclaimed here rather than smuggled in as a
+     * test-only mutator.
+     */
+    test('names the env vars it looked for when no credential exists anywhere', async () => {
+        const
+            fs           = await import('fs'),
+            os           = await import('os'),
+            path         = await import('path'),
+            shimDir      = fs.mkdtempSync(path.join(os.tmpdir(), 'neo-gh-shim-')),
+            originalPath = process.env.PATH;
+
+        fs.writeFileSync(path.join(shimDir, 'gh'), '#!/bin/sh\nexit 1\n', {mode: 0o755});
+
+        process.env.PATH = `${shimDir}${path.delimiter}${originalPath}`;
+
+        try {
+            await expect(GraphqlService.query(QUERY)).rejects.toThrow(/GH_TOKEN.*GITHUB_TOKEN/s);
+        } finally {
+            process.env.PATH = originalPath;
+            fs.rmSync(shimDir, {force: true, recursive: true});
+        }
     });
 });

--- a/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
+++ b/test/playwright/unit/ai/services/github-workflow/GraphqlService.spec.mjs
@@ -522,6 +522,92 @@ test.describe('Neo.ai.services.github-workflow.GraphqlService — credential res
      * service; that is a deliberate API cost and is left unclaimed here rather than smuggled in as a
      * test-only mutator.
      */
+    /**
+     * The cache-dependent branches, proven in an ISOLATED CHILD PROCESS.
+     *
+     * These three cannot be asserted in-worker: a successful `gh auth token` populates the service's
+     * private `#authToken`, which has no reset seam, so the credential would leak into every later
+     * test. The seam is a child process, not a production reset API — cache state cannot escape a
+     * process that exits. Lifted from the repository's existing
+     * `spawnSync(process.execPath, ['--input-type=module', '-e', …])` pattern.
+     *
+     * One child proves all three, because the ordering itself is the contract: the `gh` stub emits a
+     * DIFFERENT token per invocation, so a reused cache and a second shell-out are distinguishable.
+     *
+     *   1. call with no env       → CLI resolves, caching `cli-token-1`
+     *   2. call again, no env     → still `cli-token-1` ⇒ the cache was reused (no 2nd shell-out)
+     *   3. set `GH_TOKEN`, call   → `env-token` ⇒ env outranks a POPULATED cache
+     *
+     * Step 3 is the one the in-worker suite structurally cannot reach: every case there starts from
+     * an empty cache, so none of them can prove environment-OVER-cache rather than merely
+     * environment-when-empty.
+     */
+    test('env outranks a populated cache, and the CLI result is cached — proven in an isolated child', async () => {
+        const
+            {spawnSync} = await import('child_process'),
+            fs          = await import('fs'),
+            os          = await import('os'),
+            path        = await import('path'),
+            repoRoot    = path.resolve(import.meta.dirname, '../../../../../..'),
+            shimDir     = fs.mkdtempSync(path.join(os.tmpdir(), 'neo-gh-seq-')),
+            counter     = path.join(shimDir, 'n');
+
+        // a `gh` that returns a new token on every invocation, so cache reuse is observable
+        fs.writeFileSync(counter, '0');
+        fs.writeFileSync(path.join(shimDir, 'gh'),
+            `#!/bin/sh\nn=$(cat ${counter})\nn=$((n+1))\necho $n > ${counter}\necho cli-token-$n\n`,
+            {mode: 0o755}
+        );
+
+        const code = `
+            const root = ${JSON.stringify(repoRoot)};
+            await import(root + '/src/Neo.mjs');
+            await import(root + '/src/core/_export.mjs');
+            const {default: GraphqlService} = await import(root + '/ai/services/github-workflow/GraphqlService.mjs');
+
+            const seen = [];
+            globalThis.fetch = async (url, options) => {
+                seen.push(new Headers(options.headers).get('authorization'));
+                return new Response(JSON.stringify({data: {viewer: {login: 'x'}}}), {
+                    status : 200,
+                    headers: {'content-type': 'application/json'}
+                });
+            };
+
+            const QUERY = 'query T { viewer { login } }';
+
+            delete process.env.GH_TOKEN;
+            delete process.env.GITHUB_TOKEN;
+
+            await GraphqlService.query(QUERY);            // 1: CLI resolves + caches
+            await GraphqlService.query(QUERY);            // 2: cache reused?
+            process.env.GH_TOKEN = 'env-token';
+            await GraphqlService.query(QUERY);            // 3: env vs populated cache
+
+            console.log('RESULT ' + JSON.stringify(seen));
+        `;
+
+        const child = spawnSync(process.execPath, ['--input-type=module', '-e', code], {
+            cwd     : repoRoot,
+            encoding: 'utf8',
+            env     : {...process.env, PATH: `${shimDir}${path.delimiter}${process.env.PATH}`}
+        });
+
+        try {
+            const line = child.stdout.split('\n').find(l => l.startsWith('RESULT '));
+
+            expect(line, child.stdout + child.stderr).toBeTruthy();
+
+            const [first, second, third] = JSON.parse(line.slice('RESULT '.length));
+
+            expect(first,  'call 1 resolves through the gh CLI').toContain('cli-token-1');
+            expect(second, 'call 2 reuses the cached CLI token instead of shelling out again').toContain('cli-token-1');
+            expect(third,  'call 3 prefers GH_TOKEN over the already-populated cache').toContain('env-token');
+        } finally {
+            fs.rmSync(shimDir, {force: true, recursive: true});
+        }
+    });
+
     test('names the env vars it looked for when no credential exists anywhere', async () => {
         const
             fs           = await import('fs'),

